StreamChatClient_v3/WebSocketClient/WebSocketClient.swift

```diff
@@ -9,378 +9,257 @@ class WebSocketClient {
-//            return
-//        }
-//
-//        connectionStateAtomic.set(.reconnecting)
-//        let maxDelay: TimeInterval = min(0.5 + consecutiveFailures * 2, 25)
-//        let minDelay: TimeInterval = min(max(0.25, (consecutiveFailures - 1) * 2), 25)
-//        consecutiveFailures += 1
-//        let delay = TimeInterval.random(in: minDelay...maxDelay)
-//        logger?.log("⏳ Reconnect in \(delay) sec")
-//
-//        Timer.schedule(timeInterval: delay, queue: provider.callbackQueue) { [weak self] in
-//            self?.connectionStateAtomic.set(.notConnected)
-//            self?.connect()
-//        }
-  }
+  private func startListeningForAppStateUpdates() {
+    NotificationCenter.default.addObserver(
+      self, selector: #selector(handleAppDidEnterBackground), name: UIApplication.didEnterBackgroundNotification, object: nil
+    )
 
-  func disconnectInBackground() {
-//        provider.callbackQueue.async(execute: disconnectInBackgroundInWebSocketQueue)
+    NotificationCenter.default.addObserver(
+      self, selector: #selector(handleAppDidBecomeActive), name: UIApplication.didBecomeActiveNotification, object: nil
+    )
   }
 
-  private func disconnectInBackgroundInWebSocketQueue() {
-    guard options.contains(.stayConnectedInBackground) else {
-      disconnect(reason: "Going into background, stayConnectedInBackground is disabled")
-      return
+  @objc
+  private func handleAppDidEnterBackground() {
+    guard options.contains(.staysConnectedInBackground), connectionState.isActiveConnection else { return }
+
+    let backgroundTask = backgroundTaskScheduler.beginBackgroundTask { [weak self] in
+      self?.disconnect(source: .systemInitiated)
     }
 
     if backgroundTask != .invalid {
-      UIApplication.shared.endBackgroundTask(backgroundTask)
+      activeBackgroundTask = backgroundTask
+    } else {
+      // Can't initiate a background task, close the connection
+      disconnect(source: .systemInitiated)
     }
+  }
 
-    backgroundTask = UIApplication.shared.beginBackgroundTask { [weak self] in
-      self?.disconnect(reason: "Processing finished in background")
-      self?.backgroundTask = .invalid
-    }
+  @objc
+  private func handleAppDidBecomeActive() {
+    cancelBackgroundTaskIfNeeded()
+  }
 
-    if backgroundTask == .invalid {
-      disconnect(reason: "Can't create a background task")
+  private func cancelBackgroundTaskIfNeeded() {
+    if let backgroundTask = activeBackgroundTask {
+      backgroundTaskScheduler.endBackgroundTask(backgroundTask)
+      activeBackgroundTask = nil
     }
   }
+}
+
+extension WebSocketClient {
+  /// An object encapsulating all dependencies of `WebSocketClient`.
+  struct Environment {
+    var timer: Timer.Type = DefaultTimer.self
 
-  private func cancelBackgroundWork() {
-//        logger?.log("Cancelling background work...")
+    var notificationCenterBuilder: () -> NotificationCenter = NotificationCenter.init
```

Comment:
Why not `NotificationCenter.default`

Reply:
I didn't want to use the public one to enforce tighter coupling. You have to have a reference to `WebSocketClient` to be able to listen to new events. This notification center is internal only to LLC. Users shouldn't use it to listen for new events directly.